### PR TITLE
Dump shader programs on compile errors

### PR DESCRIPTION
**Pull request by RealBadAngel**
Example output:
```
2016-02-10 04:24:34: ERROR[Main]: Irrlicht: GLSL shader failed to compile
2016-02-10 04:24:34: ERROR[Main]: Irrlicht: 0(38) : error C1008: undefined variable "color1"

2016-02-10 04:24:34: ERROR[Main]: generate_shader(): failed to generate "selection_shader", addHighLevelShaderMaterial failed.
2016-02-10 04:24:34: ERROR[Main]: Vertex shader program:
2016-02-10 04:24:34: ERROR[Main]: ----------------------------------
2016-02-10 04:24:34: ERROR[Main]: 1: #version 120
2016-02-10 04:24:34: ERROR[Main]: 2: #define NDT_NORMAL 0
2016-02-10 04:24:34: ERROR[Main]: 3: #define NDT_AIRLIKE 1
2016-02-10 04:24:34: ERROR[Main]: 4: #define NDT_LIQUID 2
2016-02-10 04:24:34: ERROR[Main]: 5: #define NDT_FLOWINGLIQUID 3
2016-02-10 04:24:34: ERROR[Main]: 6: #define NDT_GLASSLIKE 4
2016-02-10 04:24:34: ERROR[Main]: 7: #define NDT_ALLFACES 5
2016-02-10 04:24:34: ERROR[Main]: 8: #define NDT_ALLFACES_OPTIONAL 6
2016-02-10 04:24:34: ERROR[Main]: 9: #define NDT_TORCHLIKE 7
2016-02-10 04:24:34: ERROR[Main]: 10: #define NDT_SIGNLIKE 8
2016-02-10 04:24:34: ERROR[Main]: 11: #define NDT_PLANTLIKE 9
2016-02-10 04:24:34: ERROR[Main]: 12: #define NDT_FENCELIKE 10
2016-02-10 04:24:34: ERROR[Main]: 13: #define NDT_RAILLIKE 11
2016-02-10 04:24:34: ERROR[Main]: 14: #define NDT_NODEBOX 12
2016-02-10 04:24:34: ERROR[Main]: 15: #define NDT_GLASSLIKE_FRAMED 13
2016-02-10 04:24:34: ERROR[Main]: 16: #define TILE_MATERIAL_BASIC 0
2016-02-10 04:24:34: ERROR[Main]: 17: #define TILE_MATERIAL_ALPHA 1
2016-02-10 04:24:34: ERROR[Main]: 18: #define TILE_MATERIAL_LIQUID_TRANSPARENT 2
2016-02-10 04:24:34: ERROR[Main]: 19: #define TILE_MATERIAL_LIQUID_OPAQUE 3
2016-02-10 04:24:34: ERROR[Main]: 20: #define TILE_MATERIAL_WAVING_LEAVES 4
2016-02-10 04:24:34: ERROR[Main]: 21: #define TILE_MATERIAL_WAVING_PLANTS 5
2016-02-10 04:24:34: ERROR[Main]: 22: #define MATERIAL_TYPE 1
2016-02-10 04:24:34: ERROR[Main]: 23: #define DRAW_TYPE 1
2016-02-10 04:24:34: ERROR[Main]: 24: #define GENERATE_NORMALMAPS 0
2016-02-10 04:24:34: ERROR[Main]: 25: #define NORMALMAPS_STRENGTH 0.6
2016-02-10 04:24:34: ERROR[Main]: 26: #define SAMPLE_STEP 0.00390625
2016-02-10 04:24:34: ERROR[Main]: 27: #define USE_NORMALMAPS 0
2016-02-10 04:24:34: ERROR[Main]: 28: #define ENABLE_WAVING_WATER 0
2016-02-10 04:24:34: ERROR[Main]: 29: #define ENABLE_WAVING_LEAVES 0
2016-02-10 04:24:34: ERROR[Main]: 30: #define ENABLE_WAVING_PLANTS 0
2016-02-10 04:24:34: ERROR[Main]: 31: uniform mat4 mWorldViewProj;
2016-02-10 04:24:34: ERROR[Main]: 32: 
2016-02-10 04:24:34: ERROR[Main]: 33: void main(void)
2016-02-10 04:24:34: ERROR[Main]: 34: {
2016-02-10 04:24:34: ERROR[Main]: 35: 	gl_TexCoord[0] = gl_MultiTexCoord0;
2016-02-10 04:24:34: ERROR[Main]: 36: 	gl_Position = mWorldViewProj * gl_Vertex;
2016-02-10 04:24:34: ERROR[Main]: 37: 
2016-02-10 04:24:34: ERROR[Main]: 38: 	gl_FrontColor = gl_BackColor = gl_Color;
2016-02-10 04:24:34: ERROR[Main]: 39: }
2016-02-10 04:24:34: ERROR[Main]: 40: 
2016-02-10 04:24:34: ERROR[Main]: End of Vertex shader program.
2016-02-10 04:24:34: ERROR[Main]:  
2016-02-10 04:24:34: ERROR[Main]: Pixel shader program:
2016-02-10 04:24:34: ERROR[Main]: ----------------------------------
2016-02-10 04:24:34: ERROR[Main]: 1: #version 120
2016-02-10 04:24:34: ERROR[Main]: 2: #define NDT_NORMAL 0
2016-02-10 04:24:34: ERROR[Main]: 3: #define NDT_AIRLIKE 1
2016-02-10 04:24:34: ERROR[Main]: 4: #define NDT_LIQUID 2
2016-02-10 04:24:34: ERROR[Main]: 5: #define NDT_FLOWINGLIQUID 3
2016-02-10 04:24:34: ERROR[Main]: 6: #define NDT_GLASSLIKE 4
2016-02-10 04:24:34: ERROR[Main]: 7: #define NDT_ALLFACES 5
2016-02-10 04:24:34: ERROR[Main]: 8: #define NDT_ALLFACES_OPTIONAL 6
2016-02-10 04:24:34: ERROR[Main]: 9: #define NDT_TORCHLIKE 7
2016-02-10 04:24:34: ERROR[Main]: 10: #define NDT_SIGNLIKE 8
2016-02-10 04:24:34: ERROR[Main]: 11: #define NDT_PLANTLIKE 9
2016-02-10 04:24:34: ERROR[Main]: 12: #define NDT_FENCELIKE 10
2016-02-10 04:24:34: ERROR[Main]: 13: #define NDT_RAILLIKE 11
2016-02-10 04:24:34: ERROR[Main]: 14: #define NDT_NODEBOX 12
2016-02-10 04:24:34: ERROR[Main]: 15: #define NDT_GLASSLIKE_FRAMED 13
2016-02-10 04:24:34: ERROR[Main]: 16: #define TILE_MATERIAL_BASIC 0
2016-02-10 04:24:34: ERROR[Main]: 17: #define TILE_MATERIAL_ALPHA 1
2016-02-10 04:24:34: ERROR[Main]: 18: #define TILE_MATERIAL_LIQUID_TRANSPARENT 2
2016-02-10 04:24:34: ERROR[Main]: 19: #define TILE_MATERIAL_LIQUID_OPAQUE 3
2016-02-10 04:24:34: ERROR[Main]: 20: #define TILE_MATERIAL_WAVING_LEAVES 4
2016-02-10 04:24:34: ERROR[Main]: 21: #define TILE_MATERIAL_WAVING_PLANTS 5
2016-02-10 04:24:34: ERROR[Main]: 22: #define MATERIAL_TYPE 1
2016-02-10 04:24:34: ERROR[Main]: 23: #define DRAW_TYPE 1
2016-02-10 04:24:34: ERROR[Main]: 24: #define GENERATE_NORMALMAPS 0
2016-02-10 04:24:34: ERROR[Main]: 25: #define NORMALMAPS_STRENGTH 0.6
2016-02-10 04:24:34: ERROR[Main]: 26: #define SAMPLE_STEP 0.00390625
2016-02-10 04:24:34: ERROR[Main]: 27: #define USE_NORMALMAPS 0
2016-02-10 04:24:34: ERROR[Main]: 28: #define ENABLE_WAVING_WATER 0
2016-02-10 04:24:34: ERROR[Main]: 29: #define ENABLE_WAVING_LEAVES 0
2016-02-10 04:24:34: ERROR[Main]: 30: #define ENABLE_WAVING_PLANTS 0
2016-02-10 04:24:34: ERROR[Main]: 31: uniform sampler2D baseTexture;
2016-02-10 04:24:34: ERROR[Main]: 32: 
2016-02-10 04:24:34: ERROR[Main]: 33: void main(void)
2016-02-10 04:24:34: ERROR[Main]: 34: {
2016-02-10 04:24:34: ERROR[Main]: 35: 	vec2 uv = gl_TexCoord[0].st;
2016-02-10 04:24:34: ERROR[Main]: 36: 	vec4 color = texture2D(baseTexture, uv);
2016-02-10 04:24:34: ERROR[Main]: 37: 	color.rgb *= gl_Color.rgb;
2016-02-10 04:24:34: ERROR[Main]: 38: 	gl_FragColor = color1;
2016-02-10 04:24:34: ERROR[Main]: 39: }
2016-02-10 04:24:34: ERROR[Main]: 40: 
2016-02-10 04:24:34: ERROR[Main]: End of Pixel shader program.
2016-02-10 04:24:34: ERROR[Main]:  
2016-02-10 04:24:34: ERROR[Main]: Geometry shader program:
2016-02-10 04:24:34: ERROR[Main]: ----------------------------------
2016-02-10 04:24:34: ERROR[Main]: 1: 
2016-02-10 04:24:34: ERROR[Main]: End of Geometry shader program.
2016-02-10 04:24:34: ERROR[Main]:  
```

### CODE REVIEW

File: src/shader.cpp [hunk 1]
```diff
@@ -871,3 +877,18 @@ void load_shaders(std::string name, SourceShaderCache *sourcecache,
 	}
 
 }
+
+void dumpShaderProgram(const std::string &program_type, const std::string &program)
+{
+	errorstream<<program_type<<" shader program:"<<std::endl<<
```
Comment (est31) [inline src/shader.cpp]: Write the shader program to warningstream instead. errorstream gets printed into the chat in the client and must be very concise.

File: src/shader.cpp [hunk 2]
```diff
@@ -871,3 +877,18 @@ void load_shaders(std::string name, SourceShaderCache *sourcecache,
...
+		"----------------------------------"<<std::endl;
+	size_t pos = 0;
+    size_t prev = 0;
+    s16 line = 1;
+    while ((pos = program.find("\n", prev)) != std::string::npos) {
+        errorstream<<line++<<": "<<program.substr(prev, pos - prev)<<std::endl;
```
Comment (est31) [inline src/shader.cpp]: Spaces before and after the `<<` required.

**Comment by est31** [comment]
Otherwise looks good.

**Comment by nerzhul** [comment]
some spaces are missing and you should use tabs for indent, not spaces

**Comment by est31** [comment]
Looks good.

**Comment by RealBadAngel** [comment]
ok, changed output to warningstream

File: src/shader.cpp [hunk 3]
```diff
@@ -826,6 +829,9 @@ ShaderInfo generate_shader(std::string name, u8 material_type, u8 drawtype,
 					"failed to generate \""<<name<<"\", "
 					"addShaderMaterial failed."
 					<<std::endl;
+			dumpShaderProgram("Vertex", vertex_program);
+			dumpShaderProgram("Pixel", pixel_program);
+			dumpShaderProgram("Geometry", geometry_program);
```
Comment (kahrl) [inline src/shader.cpp]: The geometry shader isn't passed to addShaderMaterial, so it shouldn't be printed here.

**Comment by kahrl** [comment]
:+1:

**Comment by kahrl** [comment]
f253ff9f1ab246c67ee5d3fb520c584b384ee708

**Pull request closed without merging**